plumpy/process_comms.py

```diff
@@ -158,8 +159,9 @@ async def pause_process(self, pid, msg=None):
         if msg is not None:
             message[MESSAGE_KEY] = msg
 
-        pause_future = await self._communicator.rpc_send(pid, message)
-        result = await pause_future
+        pause_future = self._communicator.rpc_send(pid, message)
+        future = await asyncio.wrap_future(pause_future)
+        result = await asyncio.wrap_future(future)
```

Comment:
Here is one of the places where ThreadFuture returned from communicator is convert to awaitable one and being awaited. I am totally lost in counting how many times I should unwrap the result from the future. Could you give me some hint about when should I wait for the result? @muhrin

Reply:
This code should read:
```
pause_future = await self._communicator.rpc_send(pid, message)
result = await asyncio.wrap_future(pause_future)
return result
```
The reason for this is that `await self._communicator.rpc_send(pid, message)` is just awaiting for the asynchronous communication to happen, which then returns a concurrent future whose result is the outcome of the pausing request.  You correctly convert this to an asyncio future so once you're awaited this the coroutine is ready to return the result.

Reply:
But here the `self._communicator` is the thread one (as you mentioned above, we always use thread communicator to separate out the communicator's event loop). So the `rpc_send` also return a thread future here and it is not a coroutine so can not be await.  @muhrin

Reply:
Ah yes, you're quite right!  So your original code is correct.  As we've seen this is mega confusing so would you mind adding a comment to the class explaining what's going on here.  Namely that the communicator gives back `concurrent.futures.Future`s which we convert to `asyncio.Future`s so we can await them without blocking and that, in addition, certain remote procedures themselves return furutures if they can't action the request straight away which will require another wrapping and awaiting :)

Finally, I notice that there could be a bug here (that was already there) because the RPC call that's made ends up in `Process.pause()` which, if you look at, can return a boolean if the process is terminated or already paused.  Only if that's not the case will it return a future.
Don't worry about this for this PR but I'm just noting it down as we may need to put in an `isinstance()` check before doing the second await.

Reply:
> Finally, I notice that there could be a bug here (that was already there) because the RPC call that's made ends up in Process.pause() which, if you look at, can return a boolean if the process is terminated or already paused. Only if that's not the case will it return a future.

Ha! I was just confused about the problem and trying to figure it out. Thanks for your explain.

> would you mind adding a comment to the class explaining what's going on here. 

:ok_hand: